air/src/execution_step/boxed_value/stream/stream_definition.rs

```diff
@@ -0,0 +1,553 @@
+        match generation {
+            Generation::Previous(previous_gen) => self.previous_values.add_value_to_generation(value, previous_gen),
+            Generation::Current(current_gen) => self.current_values.add_value_to_generation(value, current_gen),
+            Generation::New => self.new_values.add_to_last_generation(value),
+        }
+    }
+}
+
+impl<'value, T: 'value + TracePosOperate + fmt::Display> Stream<T> {
+    /// Removes empty generations updating data.
+    pub(crate) fn compactify(&mut self, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
+        self.previous_values.remove_empty_generations();
+        self.current_values.remove_empty_generations();
+        self.new_values.remove_empty_generations();
+
+        let start_idx = 0.into();
+        Self::update_generations(self.previous_values.slice_iter(0.into()), start_idx, trace_ctx)?;
+
+        let start_idx = self.previous_values.len();
+        Self::update_generations(self.current_values.slice_iter(0.into()), start_idx, trace_ctx)?;
+
+        let start_idx = start_idx.checked_add(self.current_values.len()).unwrap();
+        Self::update_generations(self.new_values.slice_iter(0.into()), start_idx, trace_ctx)?;
+
+        Ok(())
+    }
+
+    fn update_generations(
+        values: impl Iterator<Item = &'value [T]>,
+        start_idx: GenerationIdx,
+        trace_ctx: &mut TraceHandler,
+    ) -> ExecutionResult<()> {
+        use crate::execution_step::errors::UncatchableError;
+        use crate::execution_step::ExecutionError;
+
+        for (position, values) in values.enumerate() {
+            let generation = start_idx.checked_add(position.into()).unwrap();
+            for value in values.iter() {
+                trace_ctx
+                    .update_generation(value.get_trace_pos(), generation)
+                    .map_err(|e| ExecutionError::Uncatchable(UncatchableError::GenerationCompactificationError(e)))?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<T> Default for Stream<T> {
+    fn default() -> Self {
+        Self {
+            previous_values: <_>::default(),
+            current_values: <_>::default(),
+            new_values: <_>::default(),
+        }
+    }
+}
+
+use air_trace_handler::merger::MetApResult;
+use air_trace_handler::merger::ValueSource;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Generation {
+    Previous(GenerationIdx),
+    Current(GenerationIdx),
+    New,
+}
+
+impl Generation {
+    #[cfg(test)]
+    pub fn previous(generation_id: u32) -> Self {
+        use std::convert::TryFrom;
+
+        let generation_id = usize::try_from(generation_id).unwrap();
+        let generation_idx = GenerationIdx::from(generation_id);
+        Self::Previous(generation_idx)
+    }
+
+    #[cfg(test)]
+    pub fn current(generation_id: u32) -> Self {
+        use std::convert::TryFrom;
+
+        let generation_id = usize::try_from(generation_id).unwrap();
```

Comment:
ditto

Reply:
I'm not sure because it's worth nothing to change the argument type to `u64`.